### ISSUE TITLE
Fog::Mode.reload behaviour has changed!

### ISSUE THREAD

**Issue by gildub** [fog/fog-core#236]
The following used to work with fog-core 1.40:

https://github.com/fog/fog-openstack/blob/master/spec/image_v2_spec.rb#L142
https://github.com/fog/fog-openstack/blob/master/spec/network_spec.rb#L40

The issue seems to be that new attributes are not merged properly.

One work around:        
```ruby
def reload
    requires :identity

    object = collection.get(identity)

    return unless object

    merge_attributes(object.attributes)
    merge_attributes(object.all_associations)

    self
    rescue Excon::Errors::SocketError
    nil
end
```

**Comment by geemus** [comment]
Hmm. I would expect it still to work, at least in that we haven't really seen other reports around this.

I believe that:

`merge_attributes(object.all_associations_and_attributes)`

Should be equivalent to the two merge actions listed there, so I'm not sure where the discrepancy is sneaking in. @plribeiro3000 any chance you see something I'm missing? Thanks!

**Comment by plribeiro3000** [comment]
I would expect the same @geemus 

For what i can see in the report, manually calling one after the other (`attributes` and `all_associations`) works but not the method the returns both (`all_associations_and_attributes`).

I will take a look and see what i can find.

**Comment by plribeiro3000** [comment]
`all_associations_and_attributes` just do a merge on both hashes:

```ruby
def all_associations_and_attributes
  all_attributes.merge(all_associations)
end
```

And looking at the actual implementation of `reload`:

```ruby
def reload
  requires :identity

  object = collection.get(identity)

  return unless object

  merge_attributes(object.all_associations_and_attributes)

  self
rescue Excon::Errors::SocketError
  nil
end
```

and the workaround by @gildub i believe it is related to how `all_attributes` and `all_associations` works.

```ruby
def all_attributes
  self.class.attributes.reduce({}) do |hash, attribute|
    hash[masks[attribute]] = send(attribute)
    hash
  end
end


def all_associations
  self.class.associations.keys.reduce({}) do |hash, association|
    hash[masks[association]] = associations[association] || send(association)
    hash
  end
end
```

The just call the method definitions without any type of cache so i believe it can only be related to the method setter and getter.

If we look at the associations, they don't do any kind of cache neither conditional setting so i believe your problem is related to `attributes` and not `associations`.

Going further:

```ruby
def merge_attributes(new_attributes = {})
  new_attributes.each_pair do |key, value|
    next if self.class.ignored_attributes.include?(key)
    if self.class.aliases[key]
      send("#{self.class.aliases[key]}=", value)
    elsif self.respond_to?("#{key}=", true)
      send("#{key}=", value)
    else
      attributes[key] = value
    end
  end
  self
end
```

`merge_attributes` does not do any kind of cache either and `attributes` is as simples as:

```ruby
def attributes
  @attributes ||= {}
end
```

So i can't really see whats wrong here.

Looking at the first code line you referenced i've found [this](https://github.com/fog/fog-openstack/blob/c11e002a24a28fe14b1160dd718ea50941c5e056/lib/fog/image/openstack/v2/models/image.rb#L44-L54):

```ruby
def method_missing(method_sym, *arguments, &block)
  if attributes.key?(method_sym)
    attributes[method_sym]
  elsif attributes.key?(method_sym.to_s)
    attributes[method_sym.to_s]
  elsif method_sym.to_s.end_with?('=')
    attributes[method_sym.to_s.gsub(/=$/, '')] = arguments[0]
  else
    super
  end
end
```

which suggests me you are saving some keys using a symbol an some using a string. Maybe thats related to your problem?

And looking at the second it seems the method being called before doing a reload is [this one](https://github.com/fog/fog-openstack/blob/c11e002a24a28fe14b1160dd718ea50941c5e056/lib/fog/network/openstack/models/network.rb#L24-L27):

```ruby
def create
  merge_attributes(service.create_network(attributes).body['network'])
  self
end
```

which might also be related to why there are some differences in the reload.

I know it was working on a previous version of `fog-core` but it help if we could remove those methods and see if the error persists.

**Comment by gildub** [comment]
@plribeiro3000, I tried without success to remove the methods you mentioned. More investigation will be needed.

**Comment by gildub** [comment]
@plribeiro3000, BTW thanks for taking the time to investigate ;)

I fails the exact same way with 2.0.0
I works fine with 1.45.0
I've reversed to latter for now.

**Comment by gildub** [comment]
@plribeiro3000, it fails only for the two models mentioned.
85cfeae what effectively the one introducing

**Comment by plribeiro3000** [comment]
@gildub Since is just for this two models, i believe they are the culprit.

Since i don't have enough knowledge of `fog-openstack` could you figure out the main difference between both and the other models?

**Comment by gildub** [comment]
@plribeiro3000,

The "Image" model class is calling 'lib/fog/core/model.rb' directly, so there is nothing in between.

Actually I've found that  https://github.com/fog/fog-core/blob/master/lib/fog/core/attributes.rb#L80 is not returning the attributes that have been created dynamically.

For a given instance of Image:
```ruby
[3] pry(#<Fog::Image::OpenStack::V2::Image>)> object
=>   <Fog::Image::OpenStack::V2::Image
    id="bc3cfc5a-0c88-47b6-82c2-c6957c30d886",
    name="reloaded_image",
    visibility="private",
    tags=[],
    self="/v2/images/bc3cfc5a-0c88-47b6-82c2-c6957c30d886",
    size=nil,
    virtual_size=nil,
    disk_format=nil,
    container_format=nil,
    id="bc3cfc5a-0c88-47b6-82c2-c6957c30d886",
    checksum=nil,
    self="/v2/images/bc3cfc5a-0c88-47b6-82c2-c6957c30d886",
    file="/v2/images/bc3cfc5a-0c88-47b6-82c2-c6957c30d886/file",
    min_disk=0,
    created_at="2016-08-25T17:09:06Z",
    updated_at="2016-08-25T17:09:07Z",
    protected=false,
    status="queued",
    min_ram=0,
    owner="7b7688e5eecd4d6aa5b5b2cb3e93a778",
    properties=nil,
    metadata=nil,
    location=nil,
    network_allocated=nil,
    base_image_ref=nil,
    image_type=nil,
    instance_uuid=nil,
    user_id=nil
  >
```

```ruby
[4] pry(#<Fog::Image::OpenStack::V2::Image>)> object.attributes
=> {:status=>"queued",
 :name=>"reloaded_image",
 :tags=>[],
 :container_format=>nil,
 :created_at=>"2016-08-25T17:09:06Z",
 :size=>nil,
 :disk_format=>nil,
 :locations=>[],
 :visibility=>"private",
 :updated_at=>"2016-08-25T17:09:07Z",
 :self=>"/v2/images/bc3cfc5a-0c88-47b6-82c2-c6957c30d886",
 :owner=>"7b7688e5eecd4d6aa5b5b2cb3e93a778",
 :protected=>false,
 :min_ram=>0,
 :file=>"/v2/images/bc3cfc5a-0c88-47b6-82c2-c6957c30d886/file",
 :checksum=>nil,
 :id=>"bc3cfc5a-0c88-47b6-82c2-c6957c30d886",
 :min_disk=>0,
 :virtual_size=>nil,
 :reloadable_property=>"updated",
 :schema=>"/v2/schemas/image"}
```

`#all_attributes` is missing fields such as :reloadable_property, :schema
```ruby
[5] pry(#<Fog::Image::OpenStack::V2::Image>)> object.all_attributes
=> {:id=>"bc3cfc5a-0c88-47b6-82c2-c6957c30d886",
 :name=>"reloaded_image",
 :visibility=>"private",
 :tags=>[],
 :self=>"/v2/images/bc3cfc5a-0c88-47b6-82c2-c6957c30d886",
 :size=>nil,
 :virtual_size=>nil,
 :disk_format=>nil,
 :container_format=>nil,
 :checksum=>nil,
 :file=>"/v2/images/bc3cfc5a-0c88-47b6-82c2-c6957c30d886/file",
 :min_disk=>0,
 :created_at=>"2016-08-25T17:09:06Z",
 :updated_at=>"2016-08-25T17:09:07Z",
 :protected=>false,
 :status=>"queued",
 :min_ram=>0,
 :owner=>"7b7688e5eecd4d6aa5b5b2cb3e93a778",
 :properties=>nil,
 :metadata=>nil,
 :location=>nil,
 :network_allocated=>nil,
 :base_image_ref=>nil,
 :image_type=>nil,
 :instance_uuid=>nil,
 :user_id=>nil}
```

**Comment by plribeiro3000** [comment]
Hmm. We are getting somewhere now. 

So the problem isn't necessarily with `#all_attributes` but with the hash being different of the original one.

Can we get the api result for the last call on `#all_attributes`?

I wonder if the data being result is exactly like the one you had previously.

**Comment by plribeiro3000** [comment]
Actually, i just notice something. You don't have `reloadable_property` neither `schema` defined as attributes [here](https://github.com/fog/fog-openstack/blob/236d5667040903bb12118c89a4d78e65c6c5878f/lib/fog/image/openstack/v2/models/image.rb#L8-L42)

**Comment by plribeiro3000** [comment]
And you are setting `reloadable_property` manually here, thus why you have different hashes.

**Comment by plribeiro3000** [comment]
It all makes sense now. Your test needs a fix or either the model needs some attributes defined.

**Comment by gildub** [comment]
@plribeiro3000, yes some of those attributes are created dynamically and maybe for the purpose of testing but I'm not 100% sure if it's the only case (I didn't implement it).

So from that my question is: Do 'attributes' have to be declared as such? 

Because before fog-core 2.0.0 then didn't have to, at least from the following piece of code:
https://github.com/fog/fog-openstack/blob/236d5667040903bb12118c89a4d78e65c6c5878f/lib/fog/image/openstack/v2/models/image.rb#L50

**Comment by plribeiro3000** [comment]
Yes. Every attribute should be declared using the DSL.

I'm not sure why it was working previously and does not work now but i believe it is related.

If we do go over the `attribute` DSL:

```ruby
def attribute(name, options = {})
  type = options.fetch(:type, "default").to_s.capitalize
  Fog::Attributes.const_get(type).new(self, name, options)
end
```

It does more then just define a setter. You can typecast its value like:

```ruby
module Fog
  module Attributes
    # = Fog Timestamp Attribute
    #
    # This class handles Timestamp attributes from the providers,
    # converting Integer and String values as a real Timestamp objects
    class Timestamp < Default
      def create_setter
        model.class_eval <<-EOS, __FILE__, __LINE__
            def #{name}=(new_#{name})
              if new_#{name}.respond_to?(:to_i)
                attributes[:#{name}] = Fog::Time.at(new_#{name}.to_i)
              else
                attributes[:#{name}] = Fog::Time.parse(new_#{name}.to_s)
              end
            end
        EOS
      end
    end
  end
end
```

**Comment by plribeiro3000** [comment]
Ok. I got it now.

The problem is indeed because the attribute is not defined using the DSL.

With the inclusion of the `mask` feature, we did break it for providers that are not using the DSL.

See:

```ruby
module Fog
  module Attributes
    # = Fog Default Attribute
    #
    # This class handles the attributes without a type force.
    # The attributes returned from the provider will keep its original values.
    class Default
      attr_reader :model, :name, :squash, :aliases, :default, :as

      def initialize(model, name, options)
        .
        .
        .
        @as = options.fetch(:as, name)
        .
        .
        .
        .
        create_mask
      end

      def create_mask
        model.masks[name] = as
      end
    end
  end
end
```

So when no mask is defined using the `as` option, the mask is defaulted to the attribute itself.

Then on `Fog::Model` it does use the masks without checking if it does exists because it does believe all attributes were defined using the DSL:

```ruby
def all_attributes
  self.class.attributes.reduce({}) do |hash, attribute|
    hash[masks[attribute]] = send(attribute)
    hash
  end
end


def all_associations
  self.class.associations.keys.reduce({}) do |hash, association|
    hash[masks[association]] = associations[association] || send(association)
    hash
  end
end
```

So i believe we have 2 options here:

* Check the existence of the mask prior to use it so attributes not defined using the DSL will work
* Define all attributes using the DSL

The first option is the easiest one but it does not help in the long term where we will have providers not using the DSL because we did leave it loose.

**Comment by plribeiro3000** [comment]
Hmmmm, another options would be to add the fix on `fog-core` and add a deprecation notice in case it happens. That would be an easy fix for now with a notice for every provider not using the DSL to change it.

What do you guys think?

@geemus @gildub

**Comment by plribeiro3000** [comment]
Something like:

```ruby
def all_attributes
  self.class.attributes.reduce({}) do |hash, attribute|
    if masks[attribute].nil?
      Fog::Logger.deprecation("Please define #{attribute} using the Fog DSL")
      hash[attribute] = send(attribute)
    else
      hash[masks[attribute]] = send(attribute)
    end

    hash
  end
end


def all_associations
  self.class.associations.keys.reduce({}) do |hash, association|
    if masks[attribute].nil?
      Fog::Logger.deprecation("Please define #{attribute} using the Fog DSL")
      hash[association] = associations[association] || send(association)
    else
      hash[masks[association]] = associations[association] || send(association)
    end
    
    hash
  end
end
```

**Comment by geemus** [comment]
@plribeiro3000 got it, that clarifies a ton, thanks! Sounds good to me. Thanks for talking through this.

**Comment by gildub** [comment]
@plribeiro3000, that's great. Thanks

**Comment by plribeiro3000** [comment]
@geemus Could you release a fix please?

**Comment by geemus** [comment]
released as v2.1.1

**Comment by plribeiro3000** [comment]
Ok. I think this sounds like another misuse of the DSL.

You are defining the attribute `subnets` [here](https://github.com/fog/fog-openstack/blob/master/lib/fog/network/openstack/models/network.rb#L10):
```ruby
attribute :subnets
```

But then you are overwriting the getter defined by the DSL [here](https://github.com/fog/fog-openstack/blob/master/lib/fog/network/openstack/models/network.rb#L20-L22):
```ruby
def subnets
  service.subnets.select { |s| s.network_id == id }
end
```

Since `all_attributes` is doing a send [here](https://github.com/fog/fog-core/blob/49235a577fba083b7f3b865a9679aa523c5a10af/lib/fog/core/attributes.rb#L84):
```ruby
def all_attributes
  self.class.attributes.reduce({}) do |hash, attribute|
    if masks[attribute].nil?
      Fog::Logger.deprecation("Please define #{attribute} using the Fog DSL")
      hash[attribute] = send(attribute)
    else
      hash[masks[attribute]] = send(attribute)
    end


    hash
  end
end
```

it will trigger the GET request.

Since it used to work before we can treat this as a bug but i don't believe is the same bug.

Shall we open a new issue for this?

**Comment by gildub** [comment]
@plribeiro3000, thanks for looking into this legacy code. I've created  https://github.com/fog/fog-core/issues/244

**Comment by plribeiro3000** [comment]
@gildub 👍
